src/solidbrush.c

```diff
@@ -75,26 +74,26 @@ gdip_solidfill_setup (GpGraphics *graphics, GpBrush *brush)
 	 * brush is changed.
 	 */
 	if (solid->base.changed) {
-		A = (solid->color & 0xFF000000) >> 24;
-		R = (solid->color & 0x00FF0000) >> 16;
-		G = (solid->color & 0x0000FF00) >> 8;
-		B = (solid->color & 0x000000FF);
-
-		solid->A = (double) A / 255.0;
-		solid->R = (double) R / 255.0;
-		solid->G = (double) G / 255.0;
-		solid->B = (double) B / 255.0;
+		BYTE a = (solid->color & 0xFF000000) >> 24;
+		BYTE r = (solid->color & 0x00FF0000) >> 16;
+		BYTE g = (solid->color & 0x0000FF00) >> 8;
+		BYTE b = (solid->color & 0x000000FF);
+
+		if (a == 0) {
+			solid->A = 0;
+			solid->R = 0;
+			solid->G = 0;
+			solid->B = 0;
+		}
+		else {
+			solid->A = (double) a / 255.0;
+			solid->R = (double) r / 255.0;
+			solid->G = (double) g / 255.0;
+			solid->B = (double) b / 255.0;
+		}
```

Comment:
What is the reason for special casing `a == 0`?

The thing that I am worried about is that it may mask an issue with premultiplies vs non-premultiplied surfaces. It is probably correct, but it would be nice to have tests for some less special alpha values, such as 0x80.

Reply:
This came up when writing tests - basically if the alpha value is zero, the pixel is set to `0x00000000`. I've been thinking a little about this and taken a look at GDI+. In fact look like GDI+ premultiplies the color of the solid brush regardless of whether the surface format is premultiplied or not... But I'm not really sure how this would integrate with Cairo, considering that cairo doesn't really support premultiplied formats immediately. Also, I may be wrong but I **think** that GDI+ is all premultiplied deep in its implementation (from taking a look at the decompilation).

![screen shot 2018-09-19 at 10 34 03](https://user-images.githubusercontent.com/1275900/45744916-94a5a500-bbf7-11e8-8300-26807b44ad14.png)


Anyway: I've special cased these tests for GDI+/libgdiplus but haven't really taken an issue because the colors are so close the difference is not even noticeable. We already have a test for less special alpha numbers: e.g.
```cpp
    // Fill with brush - CompositingModeSourceCopy.
    status = GdipSetSolidFillColor (brush, 0xDDA03040);
    assertEqualInt (status, Ok);

    status = GdipSetCompositingMode (graphics, CompositingModeSourceCopy);
    assertEqualInt (status, Ok);

    status = GdipFillRectangle (graphics, brush, 0, 0, 1, 1);
    assertEqualInt (status, Ok);

    status = GdipBitmapGetPixel (bitmap, 0, 0, &color);
    assertEqualInt (status, Ok);
#if defined(USE_WINDOWS_GDIPLUS)
    assertEqualARGB (color, 0xFF8B2A37);
#else
    assertEqualARGB (color, 0xFF8B2937);
#endif
```

as you can see the colors are literally off by like 1 byte in the blue channel

Reply:
However, that is for surfaces without an alpha channel. With an alpha channel, the difference is greater. E.g.:

```cpp
	// PixelFormat32bppARGB.
    status = GdipCreateBitmapFromScan0 (256, 256, 0, PixelFormat32bppARGB, NULL, &bitmap);
    assertEqualInt (status, Ok);

    status = GdipGetImageGraphicsContext ((GpImage *) bitmap, &graphics);
    assertEqualInt (status, Ok);

    status = GdipBitmapSetPixel (bitmap, 0, 0, 0x00000000);
    assertEqualInt (status, Ok);
	
    status = GdipSetSolidFillColor (brush, 0xDDA03040);
    assertEqualInt (status, Ok);
    // Fill with brush - CompositingModeSourceCopy.
    status = GdipSetSolidFillColor (brush, 0xDDA03040);
    assertEqualInt (status, Ok);

    status = GdipSetCompositingMode (graphics, CompositingModeSourceCopy);
    assertEqualInt (status, Ok);

    status = GdipFillRectangle (graphics, brush, 0, 0, 1, 1);
    assertEqualInt (status, Ok);

    status = GdipBitmapGetPixel (bitmap, 0, 0, &color);
    assertEqualInt (status, Ok);
#if defined(USE_WINDOWS_GDIPLUS)
    assertEqualARGB (color, 0xDDA0303F);
#else
    assertEqualARGB (color, 0xDD8B2937);
#endif
```

Reply:
> But I'm not really sure how this would integrate with Cairo, considering that cairo doesn't really support premultiplied formats immediately.

Most of the Cairo _surfaces_ are actually premultiplied (at least as used by libgdiplus), which should match whatever GDI+ is doing.

I will have to think about it more though. There are way too many bugs triggered by the premultiplation done incorrectly or at wrong places. I'll add it to my backlog to investigate this further, but unfortunately I am quite swamped these days. :-/